benchmark.py

```diff
@@ -0,0 +1,212 @@
+        "client",
+        "with_gc",
+    )
+
+    def __init__(
+        self,
+        server: str,
+        concurrency: int,
+        runs: int,
+        ops_per_run: int,
+        with_gc: bool,
+    ) -> None:
+        self.server = server
+        self.concurrency = concurrency
+        self.runs = runs
+        self.ops_per_run = ops_per_run
+        self.with_gc = with_gc
+        self.client = self._build_client()
+
+    def _build_client(self) -> CacheApi:
+        if self.server:
+            try:
+                host, port_str = self.server.split(":")
+                port = int(port_str)
+            except ValueError:
+                raise ValueError("Server must be in the format <IP>:<PORT>")
+
+            connection_pool_factory_fn = connection_pool_factory_builder(
+                initial_pool_size=self.concurrency, max_pool_size=self.concurrency * 5
+            )
+        else:
+            host, port = "localhost", 11211
+            connection_pool_factory_fn = fake_connection_pool_factory_builder(
+                fake_socket=FakeSocket(response=b"VA 5 h1 f0 l6 t-1\r\nvalue\r\n"),
+                initial_pool_size=self.concurrency,
+                max_pool_size=self.concurrency * 5,
+            )
+
+        server_pool = build_server_pool(
+            servers=[ServerAddress(host, port)],
+            connection_pool_factory_fn=connection_pool_factory_fn,
+        )
+
+        pool_provider = HashRingConnectionPoolProvider(server_pool=server_pool)
+        executor = DefaultExecutor(
+            serializer=MixedSerializer(),
+            key_encoder_fn=default_key_encoder,
+            raise_on_server_error=True,
+        )
+        router = DefaultRouter(
+            pool_provider=pool_provider,
+            executor=executor,
+        )
+        return CacheClient(router=router)
+
+    def getter(self) -> None:
+        count = 0
+        for _ in range(self.runs):
+            start_time = time.time()
+            while True:
+                self.client.get(f"key{count%200}")
+                count += 1
+                if count % self.ops_per_run == 0:
+                    elapsed_time = time.time() - start_time
+                    ops_per_sec = self.ops_per_run / elapsed_time
+                    us = elapsed_time / self.ops_per_run * 1_000_000
+                    print(f"Gets: {ops_per_sec:.2f} RPS / {us:.2f} us/req")
+                    break
+
+    def run(self) -> None:
+        total = self.runs * self.ops_per_run * self.concurrency
+        print("=== Starting benchmark ===")
+        print(f" - server: {self.server or '<mocked>'}")
+        print(f" - concurrency: {self.concurrency} threads")
+        print(f" - Requests: {total/1_000_000:.2f}M")
+        print(f"    ({self.runs} runs of {self.ops_per_run} reqs per thread)")
+        print(f" - GC: {'enabled' if self.with_gc else 'disabled'}")
+        print()
+        tasks = [threading.Thread(target=self.getter) for _ in range(self.concurrency)]
```

Comment:
Why are you not using ThreadPool?

Reply:
A thread pool typically launches X threads and you give it a lot of small work to do in a queue. Here is about giving them each the same work, a busy loop. I don't see a lot of benefit on switching to a ThreadPool and I think it can greatly impact the benchmarking results.

Reply:
Well if you create a thread pool the same  size of the amount of work you're going to do the same but using an internal python tool that provides more readability and control of the join commands, that's why I asked